[#34] 사용자 프로필 조회 기능 구현

msugo1 opened the pull request:
## 📘 사용자 프로필 조회 기능 구현 전 변경사항

## 📖  Model 관련

### 📑  User

 - 사용자의 출신 국가, 현재 거주국가, 그리고 현재 거주 도시의 변수 타입을 정수로 변경

- 국가, 도시 관련 표를 따로 두고 조인해서 조회하기 위함

## 📖  DB 관련

### 📑 현재 DB에 존재하는 테이블 들의 Schema (sql 파일) 를 프로젝트에 추가

 - countries, languages, speak, towns

### 📑    테스트 클래스 변경 반영

 - User 클래스 멤버 변수 타입의 변동으로 이에 영향을 받는 클래스들 수정

 - 언어 Id 관련해서 기존에는 각 패키지마다 언어를 두어서 상수를 불러왔다면 이제는 최상위 패키지에 언어 Id 클래스를 두고 나머지 클래스에서 가져다 사용하는 방법으로 중복코드 제거

## 📘 사용자 프로필 조회 기능 구현

## 📖 Model 관련

### 📑 UserDataForProfile

 - 사용자의 정보를 담은 객체를 DB로 부터 가져오는데 사용할 클래스

### 📑 UserProfile

- 위에서 얻어온 정보와 함께 다른 여러 정보(국가, 도시, 언어 이름, 추천글) 를 조합해 `완성된 유저 프로필` 객체를 생성하는데 사용할 클래스

### 📑 LanguageDataForProfile

 - 유저 프로필에 표시할 언어 정보에 대한 객체를 DB로 부터 매핑하는데 사용할 클래스

### 📑 Language

- 유저 프로필 정보에 표시할 완성된 언어 정보(언어 이름 포함)에 대한 객체를 생성하는데  사용할 클래스

 ## 📖 Controller 계층 관련

### 📑 ProfileController

- 사용자의 프로필을 불러오는 요청을 매핑해줄 컨트롤러 클래스

 ## 📖  Service 계층 관련

### 📑 ProfileService

- 사용자의 프로필 조회 요청에 대한 비즈니스 로직을 처리할 `getUserProfile` 메소드 작성

    1. 회원 정보를 DB에서 얻어온다. (사용자가 없거나, 정보가 제대로 등록되지 않은 사용자는 조회 시 예외 발생)

    2. 프로필에 보여줄 추천 글을 `RecommendationService`을 통해 받아온다.

    3. 사용자의 프로필 정보를 담은 객체를 만들기 전, 먼저 1번에서 얻은 국가, 도시, 언어 ID 정보를 가지고 각각에 해당하는 이름을 찾고, 그 이름들을 프로필 정보를 만드는데 활용한다.

 - 해당 사용자가 가지고 있는 ID 정보에 맞는 이름을 찾을 메소드인 `matchCountry`, `matchTown`, `matchLanguages` 메소드 추가

### 📑 FetchNameService

 - DB에 따로 저장되어 있는 국가, 도시, 언어 이름에 대한 정보를 가져오는 로직을 담당할 비즈니스 계층

 - 위의 정보는 거의 바뀌지 않는 정보에 해당하므로 복잡한 조인 로직을 없애고, DB 커넥션 비용을 최대한 줄이기 위해서 한 번 가져온 데이터는 로컬 캐시에 보관해둔다.

 - `loadAllCountriesMap`, `loadAllTownsMap`, `loadAllLanguagesMap`은 각각 데이터를 가져오고 이를 캐싱하는 역할을 담당하는 메소드

 - DB에서 막 가져온 raw 데이터를 알맞은 형태 `Map<Integer (id), String (name)>으로 변경해줄 헬퍼 메소드인 `convertIntoMap` 메소드 추가

### 📑 RecommendationService

  - 사용자의 프로필에 담을 추천 글을 가져오는 로직을 담당할 `getRecommendationsByUserId` 메소드 추가

 ## 📖 Configuartion 관련

### 📑 LocalCacheConfig

 - 언어, 국가, 도시 정보를 캐시에 담기 위한 로컬 캐시에 대한 설정을 담은  클래스

  - 위의 정보는 거의 변하지 않으므로 만료 기간을 `12시간`으로 설정

 - `@EnableCaching` 어노테이션을 붙임으로써 캐시 활성화

 ## 📖 Mapper 계층 관련

### 📑ProfileMapper

 - 해당 사용자에 대한 프로필 정보를 DB에서 직접 가져올 `getUserProfileData` 메소드 추가

### 📑FetchNameMapper

 - 언어, 국가, 도시에 대한 모든 id, 이름 정보를 캐싱하기 위해 DB에서 가져올 메소드인 `fetchAll-Country/Town/Langauge-IdsAndNanes` 메소드 추가

### 📑 RecommendationMapper

 - 해당 사용자의 프로필에 담을 추천글을 DB에서 가져올 `getRecommendationsByUserId` 메소드 추가

### 📑 ProfileMapper.xml

 - 위의 로직을 처리할 쿼리문 및 `resultMap` 작성

### 📑 FetchNameMapper.xml

- 위의 로직을 처리할 쿼리문 작성

### 📑 RecommendationMapper.xml

 - 위의 로직을 처리할 쿼리문 작성

  ## 📖 상수 관련

### 📑 CacheNames

- 여러 캐시에 대한 이름을 상수 값으로 보관할 클래스 생성

  ## 📖 추가 수정 사항

### 📑 ProfileSerivce

- getUserProfile 메소드 위에 `@Transactional(readOnly=true)`를 붙여서 슬레이브 DB를 통해 데이터를 읽어오도록 변경

### 📑 RecommendationService

 - `getRecommendationsByUserId` -> `getRecommendationsForService`로 메소드 이름 변경

 - 동일한 이유로 해당 메소드에 `@Transactional(readOnly=true)` 어노테이션 부여

### 📑 RecommendationMapper.xml

 - 프로필에 담을 추천글에 경우는 10개를 제한으로 설정

 - 전체 보기 기능은 `커서 방식의 페이징`을 이용해 다른 이슈에서 해당 유저에 대한 전체 추천글 보기 기능을 구현할 예정

Commit:
[#34] 사용자 프로필 조회 기능 구현 전 변경사항

Model 관련

* User

- 사용자의 출신 국가, 현재 거주국가, 그리고 현재 거주 도시의 변수 타입을
  정수로 변경

- 국가, 도시 관련 표를 따로 두고 조인해서 조회하기 위함

SQL

- 현재 DB에 존재하는 테이블 들의 Schema 추가

- countries, languages, speak, towns

테스트 클래스 변경 반영

- User 클래스 멤버 변수 타입의 변동으로 이에 영향을 받는 클래스들 수정

- 언어 Id 관련해서 기존에는 각 패키지마다 언어를 두어서 상수를
  불러왔다면 이제는 최상위 패키지에 언어 Id 클래스를 두고 나머지
  클래스에서 가져다 사용하는 방법으로 중복코드 제거

Commit:
[#34] 사용자 프로필 조회 기능 구현

Model 관련

* UserDataForProfile

- 사용자의 정보를 담은 객체를 DB로 부터 가져오는데 사용할 클래스

* UserProfile

- 위에서 얻어온 정보와 함께 다른 여러 정보(국가, 도시, 언어 이름, 추천글) 를 조합해 `완성된 유저 프로필` 객체를 생성하는데 사용할 클래스

* LanguageDataForProfile

- 유저 프로필에 표시할 언어 정보에 대한 객체를 DB로 부터 매핑하는데
  사용할 클래스

* Language

- 유저 프로필 정보에 표시할 완성된 언어 정보(언어 이름 포함)에 대한 객체를 생성하는데
  사용할 클래스

Controller 계층 관련

* ProfileController

- 사용자의 프로필을 불러오는 요청을 매핑해줄 컨트롤러 클래스

Service 계층 관련

* ProfileService

- 사용자의 프로필 조회 요청에 대한 비즈니스 로직을 처리할
  `getUserProfile` 메소드 작성

1. 회원 정보를 DB에서 얻어온다. (사용자가 없거나, 정보가 제대로 등록되지
   않은 사용자는 조회 시 예외 발생)

2. 프로필에 보여줄 추천 글을 `RecommendationService`을 통해 받아온다.

3. 사용자의 프로필 정보를 담은 객체를 만들기 전, 먼저 1번에서 얻은 국가,
   도시, 언어 ID 정보를 가지고 각각에 해당하는 이름을 찾고, 그 이름들을
   프로필 정보를 만드는데 활용한다.

- 해당 사용자가 가지고 있는 ID 정보에 맞는 이름을 찾을 메소드인
  `matchCountry`, `matchTown`, `matchLanguages` 메소드 추가

* FetchNameService

- DB에 따로 저장되어 있는 국가, 도시, 언어 이름에 대한 정보를 가져오는
  로직을 담당할 비즈니스 계층

- 위의 정보는 거의 바뀌지 않는 정보에 해당하므로 복잡한 조인 로직을 없애고, DB 커넥션 비용을 최대한 줄이기 위해서 한 번 가져온 데이터는 로컬 캐시에 보관해둔다.

- `loadAllCountriesMap`, `loadAllTownsMap`, `loadAllLanguagesMap`은 각각
  데이터를 가져오고 이를 캐싱하는 역할을 담당하는 메소드

- DB에서 막 가져온 raw 데이터를 알맞은 형태 `Map<Integer (id), String
  (name)>으로 변경해줄 헬퍼 메소드인 `convertIntoMap` 메소드 추가

* RecommendationService

- 사용자의 프로필에 담을 추천 글을 가져오는 로직을 담당할
  `getRecommendationsByUserId` 메소드 추가

Configuartion 관련

* LocalCacheConfig

- 언어, 국가, 도시 정보를 캐시에 담기 위한 로컬 캐시에 대한 설정을 담은
  클래스

- 위의 정보는 거의 변하지 않으므로 만료 기간을 `12시간`으로 설정

- `@EnableCaching` 어노테이션을 붙임으로써 캐시 활성화

Mapper 계층 관련

* ProfileMapper

- 해당 사용자에 대한 프로필 정보를 DB에서 직접 가져올
  `getUserProfileData` 메소드 추가

* FetchNameMapper

- 언어, 국가, 도시에 대한 모든 id, 이름 정보를 캐싱하기 위해 DB에서 가져올 메소드인 `fetchAll-Country/Town/Langauge-IdsAndNanes` 메소드 추가

* RecommendationMapper

- 해당 사용자의 프로필에 담을 추천글을 DB에서 가져올
  `getRecommendationsByUserId` 메소드 추가

* ProfileMapper.xml

- 위의 로직을 처리할 쿼리문 및 `resultMap` 작성

* FetchNameMapper.xml

- 위의 로직을 처리할 쿼리문 작성

* RecommendationMapper.xml

- 위의 로직을 처리할 쿼리문 작성

상수 관련

* CacheNames

- 여러 캐시에 대한 이름을 상수 값으로 보관할 클래스 생성

Commit:
[#34] 사용자 프로필 조회 기능 구현

* ProfileSerivce

- getUserProfile 메소드 위에 `@Transactional(readOnly=true)`를 붙여서
  슬레이브 DB를 통해 데이터를 읽어오도록 변경

* RecommendationService

- `getRecommendationsByUserId` -> `getRecommendationsForService`로
  메소드 이름 변경

- 동일한 이유로 해당 메소드에 `@Transactional(readOnly=true)` 어노테이션
  부여

* RecommendationMapper.xml

- 프로필에 담을 추천글에 경우는 10개를 제한으로 설정

- 전체 보기 기능은 `커서 방식의 페이징`을 이용해 다른 이슈에서 해당 유저에 대한 전체
  추천글 보기 기능을 구현할 예정

msugo1 commented:
1. 기능 추가가 한 브랜치에 너무 많아지는 것 같아서 `유저 프로필 캐싱` 구현 등은 다른 이슈에서 작업 하겠습니다.
2. 테스트 내용도 학습이 좀 더 필요할 것 같고 피드백 반영하면서 작업하는 것이 좀 더 나을 것 같다는 생각이 들어서 별도의 이슈를 생성하여 다른 브랜치에서 작업하려고 합니다.

msugo1 commented on src/main/java/me/soo/helloworld/controller/ProfileController.java:
```diff
@@ -0,0 +1,26 @@
+
+    @LoginRequired
+    @GetMapping("/{userId}")
+    public UserProfile getOneCountryName(@PathVariable String userId) {
```
`getUserProfile` 이라고 했어야 했는데 잘못 타이핑한 걸 지금 발견했네요. 수정해서 push하겠습니다.

Commit:
[#34] 컨트롤러 메소드 이름 수정

Commit:
[#34] 접근 제어자 및 final 키워드 부여

- 롬복 `@Value`를 활용하던 Language 클래스에 `@Value` 어노테이션을 삭제하고 빠진 `private final` 키워드 부여

f-lab-dev commented on src/main/java/me/soo/helloworld/mapper/ProfileMapper.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.Optional;
+
+@Mapper
+public interface ProfileMapper {
+    public Optional<UserDataOnProfile> getUserProfileData(String userId);
```
클래스 선언부와 메소드 선언 사이에 개행을 하나 해주면 가독성이 높아집니다~

msugo1 replied:
보통 개행 하나씩 넣어주는데 메소드가 하나 밖에 없다고 신경을 덜 썼던 것 같습니다. 바로 추가하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/model/user/UserDataOnProfile.java:
```diff
@@ -0,0 +1,32 @@
+
+    private final int livingTownId;
+
+    private List<LanguageDataForProfile> languages;
```
여기는 final이 빠져있는 이유가 있으실까요?

msugo1 replied:
처음에는 final을 부여했었는데 `ProfileMapper.xml`의 resultMap을 통해 쿼리 결과를 처리하면서 MyBatis에서 다음과 같은 예외를 발생시켰습니다. 

1. `nested exception is org.apache.ibatis.reflection.ReflectionException: Error instantiating class me.soo.helloworld.model.user.UserDataOnProfile with invalid types` 

2. `Cause: java.lang.NoSuchMethodException: me.soo.helloworld.model.user.UserDataOnProfile`

문제 원인을 찾아보니 `languages` 변수를 제외한 생성자가 필요한데 모두 final이 붙은 경우 `@RequiredArgsConstructor`에 의해서 `@AllArgsConstructor`와 같은 전체 생성자만 생성되기 때문이었습니다. 

그래서 처음에는 `@RequiredArgsConstructor`를 제거하고 `@AllArgsConstructor`와 `languages를 제외한 나머지 변수들을 포함한 생성자`를 만들어줬었는데 final로 처리된 languages 인스턴스 변수에서 컴파일 오류가 발생하여, languages에서 final을 빼고 `@RequiredArgsConstructor`을 붙여줌으로써 languages를 제외한 생성자가 만들어지게 했고, 이후 전체 생성자는 `@AllArgsConstructor`을 활용해 처리해주니 오류를 해결할 수 있었습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FetchNameService.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Cacheable(COUNTRIES_MAP)
+    public Map<Integer, String> loadAllCountriesMap() {
+        return convertIntoMap(fetchNameMapper.fetchAllCountryIdsAndNames());
```
이렇게 해서 전체를 가져오는 데이터가 몇 건인지, 용량은 어느 정도인지 고려를 해보셨나요?

msugo1 replied:
구글링해보니 ASCII 표에 있는 문자들은 `utf-8 포맷이더라도 사이즈가 1 byte`라고 하더라구요. 각 국가/도시/언어 이름을 기본적으로 알파벳으로 표기하고 (1 byte), 필요할 경우만 도시 이름을 각 국의 언어로 표기 (알파벳 이외의 문자는 사이즈를 3~4 byte로 계산) 한다고 가정했습니다.

**1. 국가**
- 구글링 해본 결과 공식적으로 존재하는 국가는 `206개` 입니다. 그 중에서도 대표적인 국가 들만 추려서(대략 70개) DB에 저장해 두고 나머지는 `Others` 탭으로 넣어서 한꺼번에 묶어주면 메모리를 조금이라도 더 절약할 수 있으리라 생각했습니다.
- 가장 긴 국가를 기준으로 이름의 크기가 `알파벳 공백 포함 40자 정도` 되었습니다. 최악의 상황을 기준으로 담는 모든 국가의 이름 길이가 40자라고 가정하고, 전체 206가 국개를 담는다고 해도 `(206 * 40) / 1024 = 약 8KB`가 나왔습니다.
- 그래서 이 정도면 전체를 캐싱하고, 만료 시간을 12시간 ~ 1일 정도로 설정한 뒤, 혹시나 국가이름 관련해서 정책 상 추가 사항이나 사용자로부터 건의 사항이 들어오면 DB에 이름을 추가하고 그 다음 번 캐싱에 반영해 줄 수 있을거라 예상해 보았습니다.

**2. 도시**
- 도시의 경우 개수/용량 예상이 좀 까다로웠습니다. 국가 이름 처럼 주요 도시이름은 DB에 넣어두고 나머지는 Others로 처리해주는 방법을 따라가기로 했습니다. 한 국가당 담을 도시의 개수는 미국을 기준으로 했습니다. 땅의 크기가 커서 주요 도시도 많은 만큼 최악을 생각했을 때 좋은 기준이라 생각했습니다.  
- 구글링 해보니 `310 cities are considered at least medium cities with populations of 100,000 or more` 이라고 나와 있어서 `대략 300개`를 도시 개수 기준으로 잡았습니다. 
- 가장 긴 도시 이름이 `약 20자`를 차지한다고 하고, DB에 등록된 국가 개수를 약 100개라고 가정했을 때, `(20 * 100 * 300) / 1024 = 약 585KB`로 꽤나 크기가 크게 나오긴 했지만, 도시의 개수가 미국처럼 많을 만한 나라는 흔치 않기 때문에 각 국의 도시 이름을 자국의 언어로 표기하더라도 차지하는 메모리가 비슷하거나, 좀 더 낮을 것이라 예측했습니다.
- 때문에 현재는 전체 캐싱을 해주었으나  혹시나 차지하는 용량이 부담된다 싶으면 `국가 ID를 활용`해서 DB에서 WHERE문을 활용해 `캐싱된 국가 만의 도시들만` 가져오는 방법도 활용할 수 있을 것 같습니다.

**3. 언어**
- 언어의 수도 구글링 결과 전체는 약 7000 개 정도 입니다만 학습자들이 찾아서 학습할 만한 언어는 정해져 있습니다. 따라서 넉넉하게 약 400 개의 언어를 DB에 등록하고 나머지는 Others로 처리한다고 가정했습니다.
- 가장 긴 언어의 길이를 20자라고 가정했을 때 차지할 예상 메모리 공간은 `(400 * 20) / 1024 = 약 8KB` 입니다. 이는 국가를 캐싱했을 때와 비슷한 수치로 전체 캐싱에도 부담이 없을 것이라 판단했습니다.

f-lab-dev replied:
이렇게 프로젝트하면서 깊게까지 고민해본것들은 위키에도 잘 정리해주세요~

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FetchNameService.java:
```diff
@@ -0,0 +1,41 @@
+
+    private Map<Integer, String> convertIntoMap(List<Map<String, Object>> idsAndNames) {
+        Map<Integer, String> convertedMap = new HashMap<>();
+        idsAndNames.forEach(idAndName -> convertedMap.put((Integer) idAndName.get(ID), (String) idAndName.get(NAME)));
```
굳이 람다식 안에서 `put`을 호출하는것보다 스트림 자체에서 `Map`을 만들 수 있습니다.
`java stream list to map` 정도로 검색해보시면 될 것 같아요~

msugo1 replied:
감사합니다. 람다 식을 통해서 바로 변환이 가능하다까지는 미처 생각이 미치지 못했네요. 검색해보고 바로 바꾸도록 하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/service/ProfileService.java:
```diff
@@ -0,0 +1,60 @@
+        List<Language> languages = new ArrayList<>();
+
+        languagesData.forEach(language ->
+                languages.add(Language.builder()
```
여기도 `list.add`를 쓰기보단 스트림에서 바로 변환이 가능합니다~ `map` 메소드를 이용해보시면 될 것 같네요

Commit:
[#34] 피드백 반영

1. ProfileMapper

- 클래스 시작 부분에 개행 추가

2. FetchNamerService

- 람다식 안에서 put을 활용해 맵으로 바꾸는 구조에서 람다식의 스트림
  자체를 이용하여 map으로 바꾸도록 변경

3. ProfileService

- 람다식 안에서 add를 활용해 리스트 요소를 추가해주는 구조에서 스트림
  자체와 `map` 식을 활용해서 리스트로 바로 바꾸도록 변경
